controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,795 @@
+				Name: alertRelabelConfigSecretName,
+			},
+			Key: alertRelabelConfigSecretKey,
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to update Prometheus: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ManagedFusionDeployment) reconcileAlertmanager() error {
+	r.Log.Info("Reconciling Alertmanager")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanager, func() error {
+		if err := r.own(r.alertmanager); err != nil {
+			return err
+		}
+
+		desired := templates.AlertmanagerTemplate.DeepCopy()
+		desired.Spec.AlertmanagerConfigSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				monLabelKey: monLabelValue,
+			},
+		}
+		r.alertmanager.Spec = desired.Spec
+		utils.AddLabel(r.alertmanager, monLabelKey, monLabelValue)
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ManagedFusionDeployment) reconcileAlertmanagerSecret() error {
+	r.Log.Info("Reconciling AlertmanagerSecret")
+	managedFusionSecretData := r.managedFusionSecret.Data
+	managedFusionDeploymentPagerDutyConfig, exist := managedFusionSecretData["pager_duty_config"]
+	if !exist {
+		return fmt.Errorf("pager_duty_config key not found in managed-fusion-agent-config secret")
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentPagerDutyConfig, r.pagerDutyConfigData); err != nil {
+		return err
+	}
+	if r.pagerDutyConfigData.ServiceKey == "" {
+		return fmt.Errorf("managed-fusion-agent-config secret contains an empty pagerDutyKey entry")
+	}
+
+	managedFusionDeploymentSMTPConfig, exist := managedFusionSecretData["smtp_config"]
+	if !exist {
+		return fmt.Errorf("smtp_config key not found in managed-fusion-agent-config secret")
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentSMTPConfig, r.smtpConfigData); err != nil {
+		return err
+	}
+	if r.smtpConfigData.Password == "" {
+		return fmt.Errorf("managed-fusion-agent-config secret contains an empty smtpPassword entry")
+	}
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerSecret, func() error {
+		if err := r.own(r.alertmanagerSecret); err != nil {
+			return err
+		}
+
+		r.alertmanagerSecret.Data = map[string][]byte{
+			amSecretPagerDutyServiceKeyKey: []byte(r.pagerDutyConfigData.ServiceKey),
+			amSecretSMTPAuthPasswordKey:    []byte(r.smtpConfigData.Password),
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (r *ManagedFusionDeployment) reconcileAlertmanagerConfig() error {
+	r.Log.Info("Reconciling AlertmanagerConfig")
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfig, func() error {
+		if err := r.own(r.alertmanagerConfig); err != nil {
+			return err
+		}
+		if err := r.get(r.alertmanagerSecret); err != nil {
+			return fmt.Errorf("Unable to get alertmanager secret: %v", err)
+		}
+		if r.smtpConfigData == nil {
+			return fmt.Errorf("smtp config doesnot contain alertmanager secret")
+		}
+		if r.pagerDutyConfigData == nil {
+			return fmt.Errorf("pagerduty config doesnot contain alertmanager secret")
+		}
```

Comment:
I think the error messages are wrong. 
Also, if you passed the unmarshalling at the start of the reconciliation you are already guaranteed to have data in `r. smtpConfigData` and `r.pagerDutyConfigData`

Reply:
Updated in latest commit